services/src/pro/datasets/external/copernicus_dataspace/sentinel2.rs

```diff
@@ -0,0 +1,573 @@
+use std::path::PathBuf;
+
+use crate::pro::datasets::external::copernicus_dataspace::stac::{
+    load_stac_items, resolve_datetime_duplicates,
+};
+use gdal::{DatasetOptions, GdalOpenFlags};
+use geoengine_datatypes::{
+    primitives::{
+        CacheTtlSeconds, RasterQueryRectangle, SpatialResolution, TimeInstance, TimeInterval,
+    },
+    spatial_reference::{SpatialReference, SpatialReferenceAuthority},
+};
+use geoengine_operators::{
+    engine::{MetaData, RasterBandDescriptor, RasterResultDescriptor},
+    source::{
+        GdalDatasetParameters, GdalLoadingInfo, GdalLoadingInfoTemporalSlice,
+        GdalLoadingInfoTemporalSliceIterator,
+    },
+    util::{
+        gdal::{gdal_open_dataset_ex, gdal_parameters_from_dataset},
+        TemporaryGdalThreadLocalConfigOptions,
+    },
+};
+
+use async_trait::async_trait;
+use snafu::{ResultExt, Snafu};
+use url::Url;
+
+use super::{
+    ids::{Sentinel2Band, Sentinel2Product, UtmZone},
+    stac::{CopernicusStacError, StacItemExt},
+};
+
+const CACHE_TTL_SECS: u32 = 60 * 60 * 24 * 30; // 30 days
+const FALLBACK_CACHE_TTL_SECS: u32 = 60 * 60 * 24; // 1 day, time to cache tiles that have no known successor
+const FALLBACK_TIME_SLICE_DURATION_MILLIS: i64 = 100; // validity of tiles that have no known successor
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)))]
+#[snafu(context(suffix(false)))] // disables default `Snafu` suffix
+pub enum CopernicusSentinel2Error {
+    CannotRetrieveStacItems {
+        source: CopernicusStacError,
+    },
+    CannotParseStacUrl {
+        source: url::ParseError,
+    },
+    CannotResolveDateTimeDuplicates {
+        source: CopernicusStacError,
+    },
+    Stac {
+        source: CopernicusStacError,
+    },
+    InvalidTimeInterval {
+        source: geoengine_datatypes::error::Error,
+    },
+    CouldNotOpenDataset {
+        source: geoengine_operators::error::Error,
+    },
+    CouldNotGenerateGdalParameters {
+        source: geoengine_operators::error::Error,
+    },
+    Join {
+        source: tokio::task::JoinError,
+    },
+    MissingAssetProductUrl {
+        source: CopernicusStacError,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub struct Sentinel2Metadata {
```

Comment:
maybe name it CopernicusSentinel2Metadata or something like this because there might be more Sentinel2 sources?

Reply:
the code is in a submodule, so it should be fine? If we have multiple sources we can use the qualified name